### PR TITLE
Improve Touch Handling

### LINKED ISSUE #217: Android: Improve touch handling
SDL's touch handling makes it impossible to get accurate timestamps on events, which breaks our velocity scrolling and causes other weird behaviour.

Since we have a custom `custom/core/SDL_android.c`, we can override the entry point and do our own processing of the touch events, e.g. attaching correct timestamps, device/finger IDs and avoiding unnecessary unit conversions.

If we implement this ourselves in Swift, we will be part-way to being independent of SDL.

### PR DESCRIPTION
**Pull request by rikner**
Fixes #217 

**Type of change:** 

## Motivation (current vs expected behavior)
SDL's touch handling makes it impossible to get accurate timestamps on events, which breaks our velocity scrolling and causes other weird behaviour.

## Please check if the PR fulfills these requirements
- [x] Self-review: I am confident this is the simplest and clearest way to achieve the expected behaviour
- [x] There are no dependencies on other PRs or I have linked dependencies through Zenhub
- [x] The commit messages are clean and understandable
- [ ] Tests for the changes have been added (for bug fixes / features)

### CODE REVIEW

**Comment by rikner** [comment]
@janekszynal you might want have a look at my changes.
I was able to correctly pass timestamp (maintaining the actual value) as `JavaLong` to the swift implementation of `onNativeTouch` by using `@_cdecl` instead of `@_silgen_name`.

File: src/main/java/org/libsdl/app/SDLOnTouchListener.kt [hunk 1]
```diff
@@ -13,25 +13,14 @@ interface SDLOnTouchListener: View.OnTouchListener {
...
-
-        fun MotionEvent.touchValues(i: Int): TouchValues {
-            return TouchValues(
-                    getPointerId(i),
-                    getX(i) / mWidth,
-                    getY(i) / mHeight,
-                    // Pressure can be > 1.0 on some devices. See getPressure(i) docs.
-                    min(this.getPressure(i), 1.0f)
-            )
-        }
```
Comment (rikner) [inline src/main/java/org/libsdl/app/SDLOnTouchListener.kt]: moved this to `private` top-level

File: Sources/UIApplication+handleSDLEvents.swift [hunk 2]
```diff
@@ -14,6 +14,14 @@ extension UIApplication {
         var e = SDL_Event()
 
         while SDL_PollEvent(&e) == 1 {
+
+            #if !os(macOS)
```
Comment (michaelknoch) [inline Sources/UIApplication+handleSDLEvents.swift]:
```suggestion
            #if os(Android)
```

File: Sources/UIApplication+handleSDLEvents.swift [hunk 3]
```diff
@@ -96,3 +104,130 @@ extension SDL_Event {
...
+                at: CGPoint(
+                    x: CGFloat(event.tfinger.x),
+                    y: CGFloat(event.tfinger.y)
+                ),
+                timestamp: event.timestampInSeconds
+            )
+
+            if
+                let firstExistingEvent = UIEvent.activeEvents.first,
+                let _ = firstExistingEvent.allTouches?.first(where: {$0.touchId == event.tfinger.fingerId} )
```
Comment (michaelknoch) [inline Sources/UIApplication+handleSDLEvents.swift]:
```suggestion
                let _ = firstExistingEvent.allTouches?.first(where: { $0.touchId == event.tfinger.fingerId })
```

File: Sources/UIApplication+handleSDLEvents.swift [hunk 4]
```diff
@@ -96,3 +104,130 @@ extension SDL_Event {
...
+            {
+                //found a matching event, adding current touch to it and returning
```
Comment (michaelknoch) [inline Sources/UIApplication+handleSDLEvents.swift]:
```suggestion
                // Found a matching event, adding current touch to it and returning
```

File: Sources/UIApplication+handleSDLEvents.swift [hunk 5]
```diff
@@ -96,3 +104,130 @@ extension SDL_Event {
...
+                firstExistingEvent.allTouches?.insert(newTouch)
+                return firstExistingEvent
+            } else {
+                //no matching event found, creating a new one
```
Comment (michaelknoch) [inline Sources/UIApplication+handleSDLEvents.swift]:
```suggestion
                // No matching event found, creating a new one
```

File: Sources/UIApplication+handleSDLEvents.swift [hunk 6]
```diff
@@ -96,3 +104,130 @@ extension SDL_Event {
...
+                return UIEvent(touch: newTouch)
+            }
+
+        case SDL_FINGERMOTION:
+            if
+                let firstExistingEvent = UIEvent.activeEvents.first,
+                let matchingTouch = firstExistingEvent.allTouches?.first(where: { $0.touchId == event.tfinger.fingerId} )
```
Comment (michaelknoch) [inline Sources/UIApplication+handleSDLEvents.swift]:
```suggestion
                let matchingTouch = firstExistingEvent.allTouches?.first(where: { $0.touchId == event.tfinger.fingerId })
```

File: Sources/UIApplication+handleSDLEvents.swift [hunk 7]
```diff
@@ -96,3 +104,130 @@ extension SDL_Event {
...
+                return firstExistingEvent
+            }
+            else {
+                return nil
+            }
+
+        case SDL_FINGERUP:
+            if
+                let firstExistingEvent = UIEvent.activeEvents.first,
+                let matchingTouch = firstExistingEvent.allTouches?.first(where: {$0.touchId == event.tfinger.fingerId} )
```
Comment (michaelknoch) [inline Sources/UIApplication+handleSDLEvents.swift]:
```suggestion
                let matchingTouch = firstExistingEvent.allTouches?.first(where: { $0.touchId == event.tfinger.fingerId })
```

File: Sources/UIApplication+handleSDLEvents.swift [hunk 8]
```diff
@@ -96,3 +104,130 @@ extension SDL_Event {
...
+}
+
+extension UIEvent: CustomStringConvertible {
+    public var description: String {
+        var listOfTouches = ""
+        if let touches = allTouches {
+            for touch in touches {
+                listOfTouches += "\(touch.description),\n"
+            }
+        }
```
Comment (michaelknoch) [inline Sources/UIApplication+handleSDLEvents.swift]: we should implement this wit a map/join

File: Sources/UIScrollView+velocity.swift [hunk 9]
```diff
@@ -20,7 +20,7 @@ extension UIScrollView {
         // Otherwise we could animate after scrolling quickly, pausing for a few seconds, then letting go
         let velocityIsLargeEnoughToDecelerate = (self.panGestureRecognizer.velocity(in: self).magnitude > 10)
 
-        let nonBoundsCheckedScrollAnimationDistance = self.weightedAverageVelocity * 0.74 // hand-tuned
+        let nonBoundsCheckedScrollAnimationDistance = self.weightedAverageVelocity * 0.4 // hand-tuned
```
[the remark this answers is not in the document]
Comment (rikner) [reply Sources/UIScrollView+velocity.swift]: I played with that value myself, and found that `0.4` seems reasonable, we still might want to fine tune it. Anyway I think it's weird that we had to change it in the first place. Not sure if this valueis just some random factor chosen by us or if it has any (physical) background (e.g. like the `animationTimeConstant` below)

**Comment by ephemer** [comment]
Note to self: the current branch causes huge slowdowns with multitouch, we may need to reorganise some code to make this efficient

File: Sources/UIApplication+handleSDLEvents.swift [hunk 10]
```diff
@@ -14,6 +14,14 @@ extension UIApplication {
         var e = SDL_Event()
 
         while SDL_PollEvent(&e) == 1 {
+
+            #if os(Android)
+            if let uiEvent = UIEvent.from(e) {
+                sendEvent(uiEvent)
+                break
+            }
+            #endif
```
Comment (michaelknoch) [inline Sources/UIApplication+handleSDLEvents.swift]: its not clear why we send send events for android only

File: Sources/UIEvent+fromSDL_Event.swift [hunk 11]
```diff
@@ -0,0 +1,129 @@
+//
+//  UIEvent+fromSDL_Event.swift
+//  UIKit
+//
+//  Created by Geordie Jay on 05.11.18.
+//  Copyright © 2018 flowkey. All rights reserved.
+//
+
+#if os(Android)
```
Comment (michaelknoch) [inline Sources/UIEvent+fromSDL_Event.swift]: is this the current pattern for android only files? We could exclude them from the iOS and mac targets

File: Sources/UIEvent+fromSDL_Event.swift [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    env: UnsafeMutablePointer<JNIEnv>,
+    view: JavaObject,
+    touchDeviceID: JavaInt,
+    pointerFingerID: JavaInt,
+    action: JavaInt,
+    x: JavaFloat,
+    y: JavaFloat,
+    pressure: JavaFloat,
+    timestamp: JavaLong
+    ) {
```
Comment (michaelknoch) [inline Sources/UIEvent+fromSDL_Event.swift]:
```suggestion
) {
```
lint

File: Sources/UIApplication+handleSDLEvents.swift (hunk 10 above)
Comment (ephemer) [reply Sources/UIApplication+handleSDLEvents.swift]: I'll write a comment, thanks

File: Sources/UIEvent+fromSDL_Event.swift (hunk 11 above)
Comment (ephemer) [reply Sources/UIEvent+fromSDL_Event.swift]: Sounds good to me, thanks

File: Sources/UIEvent+fromSDL_Event.swift (hunk 12 above)
Comment (ephemer) [reply Sources/UIEvent+fromSDL_Event.swift]: Thanks, a classic Xcode formatting fail..

**Comment by rikner** [comment]
@ephemer @janek 

closing this PR in favor of 
https://github.com/flowkey/UIKit-cross-platform/pull/282/files
and
https://github.com/flowkey/UIKit-cross-platform/pull/283/files

**Pull request closed without merging**